content/blog/2021/06/writing-your-first-policy-with-kubewarden.md

```diff
@@ -0,0 +1,554 @@
+`policy.wasm`.
+
+```console
+$ file policy.wasm
+policy.wasm: WebAssembly (wasm) binary module version 0x1 (MVP)
+```
+
+Looks about right! Now, let's go ahead and write our policy!
+
+### Policy first go
+
+Let's start by editing the `main.go` file. For the sake blog post
+brevity, I'll keep the code on the `main.go` file, so it's
+straightforward to follow. Also, I'll be skipping some other best
+practices, like TDD or even having unit tests, again for the sake of
+brevity.
+
+```go
+package main
+
+import (
+	wapc "github.com/wapc/wapc-guest-tinygo"
+	kubewarden "github.com/kubewarden/policy-sdk-go"
+)
+
+func main() {
+	wapc.RegisterFunctions(wapc.Functions{
+		"validate": func(payload []byte) ([]byte, error) {
+			return kubewarden.AcceptRequest()
+		},
+		"validate_settings": func(payload []byte) ([]byte, error) {
+			return kubewarden.AcceptSettings()
+		},
+	})
+}
+```
+
+Now we have a program, but we don't have a way to know if it does what
+we want. This instance is pretty trivial, right? It does nothing. It
+just accepts the request as if there was no policy at all. If policy
+ghosting was a thing, this would be it.
+
+On the upside, though, we have a Wasm target object that we can run
+everywhere, provided we have the right Wasm engine tooling on the
+host. How cool is that?
+
+Take into account that despite the function is named
+`AcceptRequest()`, this doesn't mean the request will be accepted, it
+just means that this policy will not reject it. Other active webhooks
+or policies could reject it though. Rejecting is final: if only one
+webhook or policy rejects the request, it will be rejected.
+
+## Dry running
+
+In order to dry run our policy we will use the [`kwctl` command-line
+tool](https://github.com/kubewarden/kwctl). You can go to the
+[releases page](https://github.com/kubewarden/kwctl/releases/), and
+fetch the latest one.
+
+Check that the tool is correctly installed:
+
+```console
+$ kwctl --version
+kwctl 0.1.1
+```
+
+You can inspect the `kwctl` commands by running `kwctl
+--help`. However, we will focus on the `run` command in this post.
+
+In order to run a policy, you need two mandatory things:
+- The policy to be executed (the `policy.wasm` file we produced
+  earlier as a first smoke test, and that we will really implement in
+  the next section of this post).
+- The request to be evaluated
+    - This is the [Kubernetes `AdmissionReview`
+  object](https://github.com/kubernetes/api/blob/v0.21.1/admission/v1beta1/types.go#L34-L42),
+  or the [`AdmissionRequest` object it
+  contains](https://github.com/kubernetes/api/blob/v0.21.1/admission/v1beta1/types.go#L45-L118). Any
+  of those will work. It has to be provided in JSON format, that is
+  how the Kubernetes API server informs the webhook, and thus, the
+  policy.
+
+Now we need a request to be evaluated. We can retrieve this request by
+auditing the API server, but I will hand you an example, given the
```

Comment:
How would "auditing the API server" work ? Link an example here.

Reply:
What I mean here is that since the goal of the post is not to how to record requests, I will just give on to the reader, so they can use it right away. In my opinion, how to record requests from the API server could be another brief blog post.

Reply:
Yes, I think we should show how to do record a request inside of another blog post. I've an idea about how improve this experience (which right now is painful for all the different policy frameworks of k8s)